twoliter/src/lock/image.rs

```diff
@@ -0,0 +1,434 @@
+use super::archive::OCIArchive;
+use super::views::ManifestListView;
+use super::Override;
+use crate::common::fs::create_dir_all;
+use crate::docker::ImageUri;
+use crate::project::{Image, Vendor};
+use anyhow::{bail, Context, Result};
+use async_trait::async_trait;
+use base64::Engine;
+use futures::{pin_mut, stream, StreamExt, TryStreamExt};
+use log::trace;
+use oci_cli_wrapper::{DockerArchitecture, ImageTool};
+use semver::Version;
+use serde::{Deserialize, Serialize};
+use sha2::Digest;
+use std::fmt::{Debug, Display, Formatter};
+use std::hash::{Hash, Hasher};
+use std::path::Path;
+use tracing::{debug, error, info, instrument};
+
+/// Represents a locked dependency on an image
+#[derive(Debug, Clone, Eq, Ord, PartialOrd, Serialize, Deserialize)]
+pub(crate) struct LockedImage {
+    /// The name of the dependency
+    pub name: String,
+    /// The version of the dependency
+    pub version: Version,
+    /// The vendor this dependency came from
+    pub vendor: String,
+    /// The resolved image uri of the dependency
+    pub source: String,
+    /// The digest of the image
+    pub digest: String,
+}
+
+impl PartialEq for LockedImage {
+    fn eq(&self, other: &Self) -> bool {
+        self.source == other.source && self.digest == other.digest
+    }
+}
+
+impl Display for LockedImage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!(
+            "{}-{}@{} ({})",
+            self.name, self.version, self.vendor, self.source,
+        ))
+    }
+}
+
+/// The hash should not contain the source to allow for collision detection
+impl Hash for LockedImage {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.version.hash(state);
+        self.vendor.hash(state);
+    }
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub(crate) struct ImageMetadata {
+    /// The name of the kit
+    #[allow(dead_code)]
+    pub name: String,
```

Comment:
I think as long as name is part of the image metadata overrides with a different names won't behave as expected.

Say I have a kit dependency at `public.ecr.aws/bottlerocket/foo:v1.0.0` and I override it with `public.ecr.aws/bottlerocket/bar:v1.0.1`, the name in the image metadata (and thus the config labels) for the former will be `foo` and the latter will be `bar`. So when I push my changes for v1.0.1 to `foo`, regardless of whether the rest of the content is the same, the manifests will never have the same digests, unless I copy the image directly from `bar` to `foo`, in which case `foo` will have invalid metadata.

Can the name field be removed? If not, it might make sense to enforce that overrides must have the same name as the thing they are overriding.

Reply:
I talked this over with @cbgbt and I don't think we need to block the PR on this. It's not an issue for the SDK since it doesn't have kit metadata, and it would require specific effort to hit this issue on kits, since we publish kits to a repository matching the name of the directory that the kit is defined in. So you'd have to rename `kits/foo` to `kits/bar` in your repository to even publish the kit to a repository that has a different name than the default